pandas/core/indexes/multi.py

```diff
@@ -3095,8 +3105,56 @@ def _update_indexer(idxr, indexer=indexer):
         # empty indexer
         if indexer is None:
             return Int64Index([])._ndarray_values
+
+        if need_sort:
+            indexer = self._reorder_indexer(seq, indexer)
+
         return indexer._ndarray_values
 
+    def _reorder_indexer(self, seq, indexer: Int64Index) -> Int64Index:
+        """
+        Reorder an indexer of a MultiIndex (self) so that the label are in the
+        same order as given in seq
+
+        Parameters
+        ----------
+        seq : label/slice/list/mask or a sequence of such
+        indexer: an Int64Index indexer of self
+
+        Returns
+        -------
+        indexer : a sorted Int64Index indexer of self ordered as seq
+        """
+        from typing import Tuple
+
+        n = len(self)
+        keys = tuple()  # type: Tuple[np.ndarray, ...]
+        # For each level of the sequence in seq, map the level codes with the
+        # order they appears in a list-like sequence
+        # This mapping is then use to reorder the indexer
+        for i, k in enumerate(seq):
+            if com.is_bool_indexer(k):
+                new_order = np.arange(n)[indexer]
+            elif is_list_like(k):
+                # Generate a map with all level codes as sorted initially
+                key_order_map = np.ones(len(self.levels[i]), dtype=np.uint64) * len(
+                    self.levels[i]
```

Comment:
I *think* that if you _sort_levels_monotonic then you don't need to reorder the indexers. is this possible?

Reply:
The levels of the given index is maybe already monotonic, but we don't want to have the order as the index, but the same order as the keys in seq.
In the test case, the index are sorted, but we wanna return them in another order.
If I don't reorder the indexers, I will have to reorder the index itself before getting the indexer.
I did try this approach at the beginnig (reindexing then getting the indexer as previously done), but this is not possible due to the variety of type that can be given as key (slice, mask, array), because you can't reindex on a slice or a mask.